### ISSUE TITLE
Personal config (~/.eslintrc) doesn't load global-installed configs/parsers/plugins

### ISSUE THREAD

**Issue by matteodisabatino** [eslint/eslint#11914]
**Tell us about your environment**

* **ESLint Version:** 6.0.1
* **Node Version:** 12.4.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "browser": true,
    "commonjs": true,
    "es6": true,
    "node": true
  },
  "extends": [
    "standard"
  ],
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": [
    "html",
    "css-in-js",
    "json"
  ],
  "rules": {
    "no-const-assign": "warn",
    "no-this-before-super": "warn",
    "no-undef": "warn",
    "no-unreachable": "warn",
    "no-unused-vars": "warn",
    "constructor-super": "warn",
    "valid-typeof": "warn"
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

I updated eslint from version 5.16.0 to 6.0.1 and, since I did it, eslint failes to start and I always receive this error: 'Failed to load config "standard" to extend from eslint 6'.

**What did you expect to happen?**

I expected that all worked fine.

**Comment by mysticatea** [comment]
Thank you for your report.

How do you run `eslint`?
As [the migration guide](https://eslint.org/docs/user-guide/migrating-to-6.0.0#package-loading-simplification) says, we have changed the plugin resolution logic. Would you check if your problem is related to that or not?

**Comment by matteodisabatino** [comment]
Hi @mysticatea, thanks for you help. It seems you gotta reason because, as I read, "With ESLint v6, plugins should always be installed locally, even if ESLint was installed globally".
So there's no way to load global plugins? It's important to me...
Anyway I use eslint along with Visual Studio Code.

**Comment by mysticatea** [comment]
I'm a VSCode user, too. If you installed `eslint` in project local, VSCode uses it. I'd like to recommend to use local installation with `package.json` because it allows us to share ESLint and plugins of the same versions with the development team, contributors, and CI system.

If you want to use global installation, use `--resolve-plugins-relative-to` option to specify the location of global-installed packages. I guess it's `npm prefix -g`.

In VSCode case, configure `eslint.options.resolvePluginsRelativeTo` option in your `settings.json` of VSCode.

**Comment by matteodisabatino** [comment]
It seems not working, also via CLI.
If I run `eslint --resolve-plugins-relative-to=/usr/local/lib/node_modules/ MY_FILE.js` I receive this error:

```
Oops! Something went wrong! :(

ESLint: 6.0.1.

ESLint couldn't find the config "standard" to extend from. Please check that the name of the config is correct.

The config "standard" was referenced from the config file in "/Users/disa/.eslintrc.json".

If you still have problems, please stop by https://gitter.im/eslint/eslint to chat with the team.
```

The error persists also if I downgrade to version 6.0.0, but I'm sure that npm prefix is correct and the path of the file is correct too.

**Comment by mysticatea** [comment]
Would you provide the result of eslint command with `--debug` option?

**Comment by matteodisabatino** [comment]
Oh sure, here the output:

```
  eslint:cli CLI args: [ '--debug', '--resolve-plugins-relative-to=/usr/local/lib/node_modules/', '/Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight/insight.controller.js' ] +0ms
  eslint:cli Running on files +6ms
  eslint:ignored-paths baseDir = "/Users/disa/Work/ISO e script" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/node_modules/*"
  eslint:ignored-paths   cooked   = "/node_modules/*" +1ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/bower_components/*"
  eslint:ignored-paths   cooked   = "/bower_components/*" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = ".*"
  eslint:ignored-paths   cooked   = ".*" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "!../"
  eslint:ignored-paths   cooked   = "!../" +0ms
  eslint:ignored-paths Looking for ignore file in /Users/disa/Work/ISO e script +0ms
  eslint:ignored-paths Could not find ignore file in cwd +1ms
  eslint:ignored-paths baseDir = "/Users/disa/Work/ISO e script" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/node_modules/*"
  eslint:ignored-paths   cooked   = "/node_modules/*" +0ms
  eslint:ignored-paths addPatternRelativeToCwd:
  eslint:ignored-paths   original = "/bower_components/*"
  eslint:ignored-paths   cooked   = "/bower_components/*" +0ms
  eslint:ignored-paths Looking for ignore file in /Users/disa/Work/ISO e script +1ms
  eslint:ignored-paths Could not find ignore file in cwd +0ms
  eslint:file-enumerator Start to iterate files: [ '/Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight/insight.controller.js' ] +0ms
  eslint:file-enumerator File: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight/insight.controller.js +2ms
  eslint:cascading-config-array-factory Load config files for /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight. +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils +1ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15. +1ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15 +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud +1ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio. +1ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131 +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches/Transmit. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches/Transmit +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library/Caches. +0ms
  eslint:config-array-factory Config file not found on /Users/disa/Library/Caches +1ms
  eslint:cascading-config-array-factory No cache found: /Users/disa/Library. +1ms
  eslint:config-array-factory Config file not found on /Users/disa/Library +0ms
  eslint:cascading-config-array-factory No cache found: /Users/disa. +0ms
  eslint:cascading-config-array-factory Stop traversing because of considered root. +0ms
  eslint:cascading-config-array-factory Loading the config file of the home directory. +0ms
  eslint:config-array-factory Loading JSON config file: /Users/disa/.eslintrc.json +1ms
  eslint:config-array-factory Config file found: /Users/disa/.eslintrc.json +1ms
  eslint:config-array-factory Loading {extends:"standard"} relative to /Users/disa/.eslintrc.json +15ms
```

Moreover if I run `eslint --debug --config=~/.eslintrc.json --resolve-plugins-relative-to=/usr/local/lib/node_modules/ MY_FILE.js`

I got this output:
```
  eslint:cli CLI args: [ '--debug', '--config=/Users/disa/.eslintrc.json', '--resolve-plugins-relative-to=/usr/local/lib/node_modules/', '/Users/disa/Library/Caches/Transmit/45D5F75B-1BED-4926-8990-CE2E61D2FEAF/35.204.250.131/home/apio/apio-cloud/public/boards/acea/15/utils/insight/insight.controller.js' ] +0ms
  eslint:cli Running on files +5ms
  eslint:config-array-factory Loading JSON config file: /Users/disa/.eslintrc.json +0ms
  eslint:config-array-factory Loading {extends:"standard"} relative to /Users/disa/.eslintrc.json +12ms
```

It let me think that all gone well I always receive the error I said before.

**Comment by mysticatea** [comment]
Thank you!

Hmm. Looks like that it failed to load `eslint-config-standard` rather than plugins.

Would you check what `node -p "require.resolve('eslint-config-standard')"` command on your home directory prints?

**Comment by matteodisabatino** [comment]
I got this error:

```
internal/modules/cjs/loader.js:626
    throw err;
    ^

Error: Cannot find module 'eslint-config-standard'
Require stack:
- /Users/disa/[eval]
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:623:15)
    at Function.resolve (internal/modules/cjs/helpers.js:21:19)
    at [eval]:1:9
    at Script.runInThisContext (vm.js:123:20)
    at Object.runInThisContext (vm.js:313:38)
    at Object.<anonymous> ([eval]-wrapper:9:26)
    at Module._compile (internal/modules/cjs/loader.js:774:30)
    at evalScript (internal/process/execution.js:80:25)
    at internal/main/eval_string.js:23:3 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/Users/disa/[eval]' ]
}
```

The error is the same if i type `node -p "require.resolve('eslint-plugin-standard')"`

**Comment by mysticatea** [comment]
OK, thank you very much. I think I found what's happened.

> More precisely, ESLint v6 resolves plugins relative to the end user’s project by default, and *always resolves shareable configs and parsers relative to the location of the config file that imports them*.
>
> https://eslint.org/docs/user-guide/migrating-to-6.0.0#package-loading-simplification

This means that `extends: "foo"` works like `require("eslint-config-foo")` in the config file since ESLint 6.0.0. But this is different behavior from 5.x (global-installed ESLint had loaded from global-installed shareable configs).

If you installed that shareable config into your home directory, I guess it works.

**Comment by matteodisabatino** [comment]
So there's no other way than install the package where the config file is? Mmmm.... so bad.... I think i rollback to eslint 5. Any way thanks for your support :-)

**Comment by mysticatea** [comment]
I understand your position. I agree that the current personal config's behavior is not convenient.

@eslint/eslint-team ESLint uses `~/.eslintrc` if there are no config files in the current project (it's called personal config). I think that it's convenient if ESLint finds global-installed shareable configs and plugins when ESLint adopted the personal config. Thoughts?

**Comment by kaicataldo** [comment]
I don’t think it’s a good idea to install and run ESLint globally and I’m a little bit hesitant about adding more options, because it encourages using this behavior.

Upgrading ESLint (even a semver-minor upgrade) can lead to more linting errors, and if someone had multiple projects on their machine they would potentially have to fix up errors in every project each time they upgrade.

If this same person collaborates with others (or runs automated testing), not defining the version in package.json can lead to mismatching results between runs.

That being said, I recognize that we already support this feature and I do think this change makes sense, given that we currently are supporting global installations.

**Comment by sgehrman** [comment]
I'm getting the same problem.  will roll back eslint

**Comment by pascalduez** [comment]
Hi,

I have a similar use-case, that worked well until 6.x

We have a mono-repo, no lerna or fancy tools.  
One base `.eslintrc` config at the repo root. 
```
 {
   "root": true,
   "extends": "@company/config"
 }
```
All peerDependencies of that config are specified and installed per packages.   
This way the packages can follow different upgrade times.  
Also most packages don't require a special config, so they don't even have an `.eslintrc`. 
If one package need some special tweaks, only adding an `.eslintrc` with the tweaks does the job. 

I think that was a simple and effective pattern. Will this definitely be impossible now?

**Comment by Js-Brecht** [comment]
I've been having this same problem.  I use a global config/install because I don't like having to set up eslint every time I start a new project.  I do understand the logic behind this change, though.  So it would be worth it to me to have project scaffold I can use that has all of my default configurations, etc.  However, before I make one, I have just started copying my global config into the local project folder, and making sure I have eslint + plugins installed there.

What I thought was strange, was that my global config is still getting merged into my local config, even after the local config has been sourced.  I fixed this by creating a new subdirectory under my dev root for the config, but it still seems strange to me that you keep walking up the tree looking for configurations after you've found the one in the project dir.  Shouldn't the local one be the "source of truth", as it were?

**Comment by not-an-aardvark** [comment]
@pascalduez It's difficult to tell without more information about how you're running ESLint. I would recommend creating a new issue or stopping by the [Gitter chat](https://gitter.im/eslint/eslint), since your issue doesn't seem related to home-directory configs.

@Js-Brecht You might want to use the [`root: true`](https://eslint.org/docs/user-guide/configuring#using-configuration-files-1) option in your local config file.

**Comment by Js-Brecht** [comment]
@not-an-aardvark Ahh, I seem to have overlooked that option.

There seems to be some module resolution issues with vscode eslint extension.  When it hits a configuration file in a parent directory, it throws an exception because it cannot find the plugins/configs that I have installed in the project, same as when operating off of a global install (I don't have a global install any longer).  Also, running "npx eslint" from the cli in the project works correctly.

Should I open an issue on [vscode-eslint](https://github.com/microsoft/vscode-eslint) for this, or is it something that belongs here?

**Comment by not-an-aardvark** [comment]
I'm not positive, but at first glance that seems like an issue with `vscode-eslint`.

**Comment by wintercounter** [comment]
We're having hundreds of projects without any node_modules folder. That is because we're using a unified dev env/stack to prevent pollutiong devs machines and have fast CI without the need to install. For this reason ESLint is also global. We need to roll back also to 5.X, because we cannot use it this way, for us local modules is just unacceptable. At least it would be nice than to be able use plugins/configs as `require.resolve('my-eslint-plugin')`, but that's not possible either, because it cannot be a path.

**Comment by JayCanuck** [comment]
Similar situation to wintercounter here and similarly this hard-locks development to eslint 5.x.

**Comment by wintercounter** [comment]
Can we get some love here please? We cannot move forward to 6.x. We cannot have local deps, not allowed by our systems and never will be. Is it really necessary to force one use case?

**Comment by not-an-aardvark** [comment]
You can make ESLint load globally-installed plugins by running it as follows:

```bash
$ eslint <yourFiles.js> --resolve-plugins-relative-to=$(npm root -g)
```

While we don't recommend this for new projects (since this makes it confusing to keep plugins up to date with your config), it can be used as a workaround to allow you to update to ESLint 6 if your system needs global installations.

**Comment by wintercounter** [comment]
Unfortunately that only work with plugins and not with config. For now
we're testing to require.resolve them but it's causing several issues at
several places.

**Comment by wintercounter** [comment]
`--resolve-plugins-relative-to` also won't work in case ESLint is being run by some 3rd party, eg: Webpack, VSCode, WebStorm, etc.
If that config would be exposed in `.eslintrc`, that would also work nicely. An additional `--resolve-configs-relative-to` would make it even better.

**Comment by scottfr** [comment]
Rolling back to 5.16 because of this issue.

For our use cases we want to have a single config automatically applied to all projects instead of individual projects being able to have different package versions and configs. Please restore the ability to have shared configs installed globally.

**Comment by wintercounter** [comment]
For us this is not a solution:
- I cannot ask all my users using my software to set NODE_PATH on their own
(many company systems doesn't even allow.)
- Cannot pass `--resolve-plugins-relative-to` to 3rd party toolings such as
`eslint-loader`.
- As mentioned before, local node_modules folder is not allowed in our case.

I don't know what happened around the house of ESLint lately but the last 2
months 3 ESLint related packages introduced such changes that totally broke
our workflows making impossible to update...

**Comment by wintercounter** [comment]
No, it doesn't make migration difficult. It makes it impossible for many of us. I know I can pass loader options. But again, it would require manual config from every user for every project then, which is unacceptable.

Imo ESLint should be able to resolve these things relative to the `eslint` module directory as well and that would solve this also.

Anyway, I'm starting to feel like I'll need to patch manually ESLint on file-system level to make this work. I really want to avoid this...

There would be so many solutions to resolve this issue:
- Allow globals again.
- Resolve configs/plugins relative to the loaded eslint module itself.
- Add support for `--resolve-configs-relative-to` flag + add solution to pass CLI options in `.eslintrc` so 3rd party tools can automatically apply without extra config.
- Allow configs to a be a file path (plugins can be, why not configs?)
- Add option to define our own `node_modules` dir(s) to resolve from. (from rc file, not as a CLI flag).

I'm sure this thread will keep growing with reports from other users when they start to update their codebase.

**Comment by kurtextrem** [comment]
A short question: If we have `--resolve-plugins-relative-to`, why not also add `--resolve-presets-relative-to`? That would literally be just a copy & paste with minimal modifications and please everyone in this thread, future and existing issues all around the ecosystem.

**Comment by tonywoode** [comment]
my 2cents: I can't upgrade eslint past v5, as ALE in Vim will report 'Failed to load config "standard"' - just as CLI does. I've read all issues here related to forcing users to stop using a global config, and I don't agree for reasons already cited above in this ticket. I see no option open to me atm than to slowly die on v5 or switch to a different linting tool?

**Comment by wintercounter** [comment]
I had enough of this and created the following package:
https://github.com/wintercounter/eslint-global-patch

It is a very unsafe and ugly workaround, it'll patch your ESLint installation. For my use case it works fine for all our projects.

**Comment by luzik** [comment]
I believe that 5.x issues "design bug" affected less installations that 6.x design broke now. Standalone ESLint (outside node project) is now quite unusable. I have to change my editor --resolve-plugins-relative-to based on that if local project use node.js.

IMHO best solutions should be:
1. add a way to include --resolve-configs-relative-to option in .eslintrc
2. detect global .eslintrc usage and as a fallback to local plugins search user global npm installation

**Comment by wintercounter** [comment]
What it needs is that is working as any other module in the NodeJS ecosystem. ESLint cannot resolve modules **relative to itself** currently!

**Comment by matteodisabatino** [comment]
I updated to eslint 6.7.2 but still I'm unable to use global modules. When i run `eslint --resolve-plugins-relative-to /usr/local/lib MYFILE.js` or `eslint --resolve-plugins-relative-to /usr/local/lib//usr/local/lib/node_modules MYFILE.js` I always receive this message:

```
Oops! Something went wrong! :(

ESLint: 6.7.2.

ESLint couldn't find the config "standard" to extend from. Please check that the name of the config is correct.

The config "standard" was referenced from the config file in "/Users/disa/.eslintrc.json".

If you still have problems, please stop by https://gitter.im/eslint/eslint to chat with the team.
```

`standard` and `eslint-plugin-standard`are installed as global and my `.eslintrc.json` is:

```json
{
  "env": {
    "browser": true,
    "commonjs": true,
    "es6": true,
    "node": true
  },
  "extends": [
    "standard"
  ],
  "parser": "babel-eslint",
  "parserOptions": {
    "ecmaFeatures": {
      "jsx": true
    },
    "sourceType": "module"
  },
  "plugins": [
    "html",
    "css-in-js",
    "json"
  ],
  "rules": {
    "no-const-assign": "warn",
    "no-prototype-builtins": "off",
    "no-this-before-super": "warn",
    "no-undef": "warn",
    "no-unreachable": "warn",
    "no-unused-vars": "warn",
    "constructor-super": "warn",
    "valid-typeof": "warn"
  }
}
```

What am I wronging now?

**Comment by devsnek** [comment]
My solution:

```js
// ~/.eslintrc.js

'use strict';
const Module = require('module');

const hacks = [
  'eslint-config-airbnb-base',
  'babel-eslint',
];

const ModuleFindPath = Module._findPath;
Module._findPath = (request, paths, isMain) => {
  const r = ModuleFindPath(request, paths, isMain);
  if (!r && hacks.includes(request)) {
    return require.resolve(`${process.env.HOME}/.npm-global/lib/node_modules/${request}`);
  }
  return r;
};

module.exports = {
  extends: 'airbnb-base',
  parser: 'babel-eslint',
  ... etc
};
```

this is so incredibly stupid. please add back global resolution in some capacity.
